include/sorting_network.h

```diff
@@ -0,0 +1,88 @@
+/*********************                                                        */
+/*! \file sorting_network.h
+** \verbatim
+** Top contributors (to current version):
+**   Makai Mann
+** This file is part of the smt-switch project.
+** Copyright (c) 2020 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief Implements a symbolic sorting network for boolean-sorted variables.
+**
+**/
+
+#pragma once
+
+#include "smt.h"
+
+namespace smt {
+
+/** \class SortingNetwork
+ *         Implements a symbolic sorting network for boolean-sorted variables.
+ *         All methods are const so it can be re-used for terms from the same
+ *         solver.
+ *         Does not return the actual terms, but is used to determine how many
+ *         variables are assigned to true symbolically.
+ *         Definition:
+ *            A sorting network takes a list of boolean terms
+ *             and returns a list of the same length.
+ *            Let the input be x := [x0, ..., xk],
+ *             then the output of a sorting network is a list
+ *             of boolean terms, y := [y0, ..., yk] such that
+ *             any model assigns yi to true iff i or more
+ *             elements of x were set to true
+ *            Thus, ignoring the edge cases, Nt = n iff
```

Comment:
Please consider not ignoring the edge cases :-) Maybe adding them after this explanation. Not critical though.

Reply:
Haha, added with 6fee1c02d9074ff5accb4d3094d3f37eb1cba863